### ISSUE TITLE
Suggestion: Use Repeat only when essential

### ISSUE THREAD

**Issue by chrysn** [core-wg/oscoap#141]
In -04 6.3.1, where the use of Repeat to synchronize the replay window is described, I suggest the following addition at the end of the paragraph:

> The server MAY defer that synchronization to the first request that is not safe (in the sense of the GET method being safe). With responses sent before synchronization has happened, the server MUST use an explicit sequence number and MUST NOT use the first-byte-flipped request sequence number.

A safe request has (by definition) no side effects, and the attacker can not gain any more information from the response than it could have gotten from the original response.

As this depends on #127, if my suggestions there are turned down, this can be closed as "won't fix".

This would also mitigate the surge of Repeat reponses in multicast cases when a new multicaster appears in some cases. (@jimsch brought that up recently)

**Comment by chrysn** [comment]
This could also go into what I'm thinking of as a "Secret Forbidden OSCOAP" nonstandard / side letter / blog post text that describes ideas that should work and co-motivate some of the laxer checking (eg. that clients must always accept both minted and bitflipped IVs), but are "don't do this at home" kind of things that are not explicitly allowed but the communication partner will accept due to some sprinkled-around MUSTs. (Does IETF have a process for such texts? Probably not, and I don't think they belong here).

**Comment by gselander** [comment]
OK, then I don't include that and close this issue.

**Comment by chrysn** [comment]
Just a brief comment as I stumbled upon that in an [german article](https://www.heise.de/ct/ausgabe/2017-4-Was-die-anstehende-Version-TLS-1-3-bringt-3608879.html) about TLS1.3 w/o having found it in the spec yet: It seems that HTTP over TLS has something very similar to this issue, the article claims that a particularly fast (0 RT) form of session resumption is only permissible if the first request is safe.

**Comment by jimsch** [comment]
The amount of argument in the TLS working group over what can "safely" go into the 0-RTT data has been long and harsh.  It is STRONGLY recommended that both the server and the client implement different APIs for handing 0-RTT data vs normal data for this reason.  This makes me leery of using 0-RTT in CoAP because of this issue and how easy it is to get wrong.  0-RTT == replay
